// @neondatabase/neon-js/dist/index.mjs
import { createInternalNeonAuth } from "@neondatabase/auth";
import { BetterAuthVanillaAdapter, SupabaseAuthAdapter } from "@neondatabase/auth/vanilla/adapters";
import "@neondatabase/auth/react/adapters";
import { AuthRequiredError, NeonPostgrestClient, fetchWithToken, fetchWithToken as fetchWithToken$1 } from "@neondatabase/postgrest-js";

//#region src/client/neon-client.ts
/**
* Neon client with integrated authentication
*
* Extends NeonPostgrestClient with Neon Auth integration.
* For auth-free clients, use @neondatabase/postgrest-js instead.
*/
var NeonClient = class extends NeonPostgrestClient {
	auth;
	constructor({ dataApiUrl, options, authClient }) {
		super({
			dataApiUrl,
			options
		});
		this.auth = authClient.adapter;
	}
};

//#endregion
//#region src/client/client-factory.ts
function createClient(config) {
	const { auth: authConfig, dataApi: dataApiConfig } = config;
	const auth = createInternalNeonAuth(authConfig.url, {
		adapter: authConfig.adapter,
		allowAnonymous: authConfig.allowAnonymous ?? false
	});
	const getAccessToken = async () => {
		return auth.getJWTToken();
	};
	const authFetch = fetchWithToken$1(getAccessToken, dataApiConfig.options?.global?.fetch);
	return new NeonClient({
		dataApiUrl: dataApiConfig.url,
		authClient: auth,
		options: {
			...dataApiConfig.options,
			global: {
				...dataApiConfig.options?.global,
				fetch: authFetch
			}
		}
	});
}

//#endregion
export { AuthRequiredError, BetterAuthVanillaAdapter, SupabaseAuthAdapter, createClient, fetchWithToken };
