// @neondatabase/postgrest-js/dist/client-BMaf7WEK.mjs
import { PostgrestClient } from "@supabase/postgrest-js";

//#region src/client/postgrest-client.ts
/**
* Neon PostgreSQL client for querying the Neon Data API
*
* This is a generic PostgreSQL client without authentication built-in.
* For auth-integrated clients, use @neondatabase/neon-js instead.
*
* Extends the upstream PostgrestClient with Neon-specific configuration.
*/
var NeonPostgrestClient = class extends PostgrestClient {
	constructor({ dataApiUrl, options }) {
		super(dataApiUrl, {
			headers: options?.global?.headers,
			fetch: options?.global?.fetch,
			schema: options?.db?.schema
		});
	}
};

//#endregion
//#region src/client/fetch-with-token.ts
/**
* Error thrown when authentication is required but no token is available
*/
var AuthRequiredError = class extends Error {
	constructor(message = "Authentication required. A valid token is needed to access the resource.") {
		super(message);
		this.name = "AuthRequiredError";
	}
};
/**
* Creates a fetch wrapper that injects authentication tokens into every request
*
* This is a generic utility that can work with any token provider function.
* The token is resolved lazily on each request.
*
* @param getAccessToken - Async function that returns current access token
* @param customFetch - Optional custom fetch implementation
* @returns Wrapped fetch function with authentication headers
*/
function fetchWithToken(getAccessToken, customFetch) {
	const baseFetch = customFetch ?? fetch;
	return async (input, init) => {
		const accessToken = await getAccessToken();
		if (!accessToken) throw new AuthRequiredError();
		const headers = new Headers(init?.headers);
		if (!headers.has("Authorization")) headers.set("Authorization", `Bearer ${accessToken}`);
		return baseFetch(input, {
			...init,
			headers
		});
	};
}

//#endregion
export { fetchWithToken as n, NeonPostgrestClient as r, AuthRequiredError as t };
